src/_P164_gases_ens160.ino

```diff
@@ -0,0 +1,158 @@
+      Device[deviceCount].VType              = Sensor_VType::SENSOR_TYPE_DUAL;
+      Device[deviceCount].Ports              = 0;
+      Device[deviceCount].PullUpOption       = false;
+      Device[deviceCount].InverseLogicOption = false;
+      Device[deviceCount].FormulaOption      = true;
+      Device[deviceCount].ValueCount         = 2;
+      Device[deviceCount].SendDataOption     = true;
+      Device[deviceCount].TimerOption        = true;
+      Device[deviceCount].GlobalSyncOption   = true;
+      Device[deviceCount].PluginStats        = true;
+      Device[deviceCount].I2CNoDeviceCheck   = true;  //TODO
+      break;
+    }
+
+    case PLUGIN_GET_DEVICENAME:
+    {
+      string = F(PLUGIN_NAME_164);
+      break;
+    }
+
+    case PLUGIN_GET_DEVICEVALUENAMES:
+    {
+      strcpy_P(ExtraTaskSettings.TaskDeviceValueNames[0], PSTR(PLUGIN_VALUENAME1_164));
+      strcpy_P(ExtraTaskSettings.TaskDeviceValueNames[1], PSTR(PLUGIN_VALUENAME2_164));
+      break;
+    }
+
+    case PLUGIN_I2C_HAS_ADDRESS:
+    case PLUGIN_WEBFORM_SHOW_I2C_PARAMS:
+    {
+      const uint8_t i2cAddressValues[] = { ENS160_I2CADDR_0, ENS160_I2CADDR_1 };
+      constexpr int nrAddressOptions   = NR_ELEMENTS(i2cAddressValues);
+
+      if (function == PLUGIN_WEBFORM_SHOW_I2C_PARAMS) {
+        addFormSelectorI2C(F("i2c_addr"), nrAddressOptions, i2cAddressValues, P164_I2C_ADDR);
+        addFormNote(F("ADDR Low=0x52, High=0x53"));
+      } else {
+        success = intArrayContains(nrAddressOptions, i2cAddressValues, event->Par1);
+      }
+
+      break;
+    }
+
+    # if FEATURE_I2C_GET_ADDRESS
+    case PLUGIN_I2C_GET_ADDRESS:
+    {
+      event->Par1 = P164_I2C_ADDR;
+      success     = true;
+      break;
+    }
+    # endif // if FEATURE_I2C_GET_ADDRESS
+
+    case PLUGIN_SET_DEFAULTS:
+    {
+      P164_I2C_ADDR = ENS160_I2CADDR_1;
+      success = true;
+      break;
+    }
+
+    case PLUGIN_INIT:
+    {
+      initPluginTaskData(event->TaskIndex, new (std::nothrow) P164_data_struct(event));
+      P164_data_struct *P164_data = static_cast<P164_data_struct *>(getPluginTaskData(event->TaskIndex));
+
+      success = (nullptr != P164_data && P164_data->begin());
+      break;
+    }
+
+    case PLUGIN_READ:
+    {
+      P164_data_struct *P164_data = static_cast<P164_data_struct *>(getPluginTaskData(event->TaskIndex));
+
+      if (nullptr == P164_data) {
+        Serial.print("P164: plugin_read NULLPTR");
+        break;
+      }
+
+      float temperature = 20.0f;  // A reasonable value in case temperature source task is invalid
+      float humidity = 50.0f;     // A reasonable value in case humidity source task is invalid
+      if (validTaskIndex(P164_PCONFIG_TEMP_TASK) && validTaskIndex(P164_PCONFIG_HUM_TASK))
+      {
+        // we're checking a var from another task, so calculate that basevar
+        temperature = UserVar[P164_PCONFIG_TEMP_TASK * VARS_PER_TASK + P164_PCONFIG_TEMP_VAL]; // in degrees C
```

Comment:
I don't see checks for the range of this array index.

Reply:
It is copied from other plugins using external temperature and humidity compensation. I assumed the validTaskIndex() will secure the range validity. If not I would like to know how to do a reasonable check.

Reply:
`validTaskIndex(P164_PCONFIG_TEMP_TASK)` is a good check.
But you can also not use the `operator[]` call of `UserVar`, but use the `getFloat()` function which needs 2 parameters (taskIndex and taskVarIndex)

I think I will later add some extra checks to make sure you're fetching the right taskvalue index as not all plugins output 4 floats. For example RFID plugins output a single 64-bit unsigned int.
But that's something for later, just some info for you to maybe better understand the code.

Reply:
I am still learning the ESPEasy internals... I like the getFloat() much better than the index juggling. The original code was copied from an old plugin.